nilearn/input_data/base_masker.py

```diff
@@ -37,7 +39,11 @@ def filter_and_mask(imgs, mask_img_,
         class_name = enclosing_scope_name(stack_level=2)
 
     mask_img_ = _utils.check_niimg(mask_img_, ensure_3d=True)
-    imgs = _utils.check_niimgs(imgs, accept_3d=True)
+
+    if sample_mask is not None:
+        imgs = image.index_img(imgs, sample_mask)
+    else:
+        imgs = _utils.check_niimgs(imgs, accept_3d=True)
```

Comment:
I suggested that because I thought that `index_img` could be faster if `imgs` was a list of 3D imgs. Given that it's not the case, I would be in favor of what @lesteve proposes.